endgame_models/get_differences.py

```diff
@@ -0,0 +1,99 @@
+from dataclasses import dataclass
+from typing import Optional, TypeVar
+
+from typing_extensions import Type
+
+from .models import BaseInitialParams
+
+Flat = str | float | int | bool
+
+JSONType = Flat | list["JSONType"] | dict[str, "JSONType"] | tuple["JSONType", ...]
+DictType = dict[str, JSONType]
+ListType = list[JSONType]
+TupleType = tuple[JSONType, ...]
+
+
+@dataclass
+class ReadOnlyDiff:
+    original: JSONType
+    new: JSONType
+
+
+OutputReadOnly = dict[str, "ReadOnlyDiff | OutputReadOnly"]
+
+
+def _output_read_only_diff(
+    initial_dict: DictType, update_dict: DictType, model: Type[BaseInitialParams]
+) -> Optional[OutputReadOnly]:
+    changed_read_only: OutputReadOnly = {}
+    for k, new_v in update_dict.items():
```

Comment:
How about this: `for k, new_v, old_v in {k, new_v, (old_v := initial_dict.get(k)) for update_dict.items() if old_v is not None}:` Feel free to ignore if you thing it's too cryptic. Removes one level of indentation.

Reply:
Yeah too obfuscated I think

Reply:
See what you were going for though